compiler/rustc_mir_transform/src/coroutine/relocate_upvars.rs

```diff
@@ -0,0 +1,324 @@
+                    source_info,
+                    kind: StatementKind::StorageDead(local),
+                }],
+                terminator: Some(Terminator { source_info, kind: TerminatorKind::Goto { target } }),
+                is_cleanup: false,
+            });
+            drop_one = patch.new_block(BasicBlockData {
+                terminator: Some(Terminator {
+                    source_info,
+                    kind: TerminatorKind::Drop { place, target: drop_one, unwind, replace },
+                }),
+                statements: vec![],
+                is_cleanup: false,
+            });
+            target = drop_one;
+            unwind = UnwindAction::Cleanup(unwind_one);
+            cleanup = unwind_one;
+        }
+    }
+    patch.patch_terminator(block, TerminatorKind::Goto { target });
+}
+
+fn rewrite_drop_coroutine_struct<'tcx>(
+    body: &mut Body<'tcx>,
+    substitution_mapping: &IndexSlice<FieldIdx, UpvarSubstitution<'tcx>>,
+) {
+    let mut blocks = vec![];
+    for (block, block_data) in body.basic_blocks.iter_enumerated() {
+        let Terminator { source_info: _, kind: TerminatorKind::Drop { place, .. } } =
+            block_data.terminator()
+        else {
+            continue;
+        };
+        let Some(local) = place.as_local() else { continue };
+        if local == ty::CAPTURE_STRUCT_LOCAL {
+            blocks.push(block)
+        }
+    }
+    let mut patch = MirPatch::new(body);
+    for block in blocks {
+        rewrite_one_drop_coroutine_struct(&mut patch, body, block, substitution_mapping);
+    }
+    patch.apply(body);
+}
+
+fn insert_substitution_prologue<'tcx>(
+    body: &mut Body<'tcx>,
+    substitution_mapping: &IndexSlice<FieldIdx, UpvarSubstitution<'tcx>>,
+) {
+    let mut patch = MirPatch::new(body);
+    let mut stmts = Vec::with_capacity(2 * substitution_mapping.len());
+    for &UpvarSubstitution { local, place, upvar_place, span } in substitution_mapping {
+        let source_info = SourceInfo::outermost(span);
+        stmts.push(Statement { source_info, kind: StatementKind::StorageLive(local) });
+        stmts.push(Statement {
+            source_info,
+            kind: StatementKind::Assign(Box::new((place, Rvalue::Use(Operand::Move(upvar_place))))),
+        });
+    }
+    let source_info = SourceInfo::outermost(body.span);
+    let resume_block = patch.resume_block();
+    let prologue = patch.new_block(BasicBlockData {
+        statements: stmts,
+        terminator: Some(Terminator {
+            source_info,
+            kind: TerminatorKind::Drop {
+                place: Place::from(ty::CAPTURE_STRUCT_LOCAL),
+                target: START_BLOCK,
+                unwind: UnwindAction::Cleanup(resume_block),
+                replace: false,
+            },
+        }),
+        is_cleanup: false,
+    });
+    patch.apply(body);
+
+    // Manually patch so that prologue is the new entry
+    let preds = body.basic_blocks.predecessors()[START_BLOCK].clone();
```

Comment:
Nit: `predecessors` is cached. It may be interesting to chose the order in which we do things, to avoid clearing the cache just before trying to access it.